R/bootstrap-model.R

```diff
@@ -0,0 +1,668 @@
+    mod_names <- purrr::map_chr(n_seq, max_char = nchar(n), pad_left)
+    mod_paths <- file.path(boot_dir, mod_names)
+    orig_mod <- read_model(get_based_on(.boot_run))
+
+    # TODO: only include subjects that entered the original problem
+    starting_data <- nm_data(orig_mod) %>% suppressMessages()
+    if(!is.null(strat_cols)){
+      checkmate::assert_true(strat_cols %in% names(starting_data))
+    }
+
+    boot_args <- list(
+      boot_run = .boot_run,
+      all_mod_names = mod_names,
+      orig_mod_path = get_model_path(orig_mod),
+      orig_mod_id = get_model_id(orig_mod),
+      orig_data = starting_data,
+      strat_cols = strat_cols,
+      seed = seed,
+      n_samples = n,
+      boot_dir = boot_dir,
+      boot_data_dir = boot_data_dir,
+      overwrite = .overwrite
+    )
+
+    # Create model object per boot run
+    boot_models <- purrr::map(mod_paths, make_boot_run, boot_args)
+    make_boot_spec(boot_models, boot_args)
+
+    # Garbage collect - may help after handling many (potentially large) datasets
+    #  - It can be useful to call gc() after a large object has been removed, as
+    #    this may prompt R to return memory to the operating system.
+    gc()
+  }else{
+    rlang::abort(
+      c(
+        glue("Bootstrap run has already been set up at `{boot_dir}`"),
+        "pass `.overwrite = TRUE` to overwrite"
+      )
+    )
+  }
+  return(invisible(.boot_run))
+}
+
+
+#' Set up a single bootstrap model run
+#'
+#' @param mod_path absolute model path (no file extension) of a bootstrap model run.
+#' @param boot_args list of parameters needed to create a bootstrap model run.
+#'
+#' @keywords internal
+make_boot_run <- function(mod_path, boot_args){
+
+  # Update the user every 100 runs (plus beginning and end)
+  new_mod_index <- which(basename(mod_path) == boot_args$all_mod_names)
+  if(new_mod_index == 1 || new_mod_index == length(boot_args$all_mod_names) ||
+     new_mod_index %% 100 == 0){
+    verbose_msg(glue("Sampling run {new_mod_index}/{length(boot_args$all_mod_names)}"))
+  }
+
+  # Copy over control stream
+  #  - Cant use copy_model_from, as we want these individual model runs to be
+  #    regular `bbi_base_model` objects
+  orig_mod_path <- boot_args$orig_mod_path
+  mod_path_ext <- paste0(mod_path, ".", fs::path_ext(orig_mod_path))
+  fs::file_copy(
+    orig_mod_path, mod_path_ext, overwrite = boot_args$overwrite
+  )
+
+
+  # Sample data and assign new IDs
+  data_new <- resample_df(
+    boot_args$orig_data,
+    key_cols = "ID", # TODO: should this be a user arg?
+    strat_cols = boot_args$strat_cols,
+    replace = TRUE
+  ) %>% dplyr::rename("OID" = "ID", "ID" = "KEY") %>% # TODO: should this be a user arg?
+    dplyr::select(all_of(unique(c(names(boot_args$orig_data), "OID")))) # TODO: should this be a user arg?
+
+  mod_name <- basename(mod_path)
+  orig_mod_id <- boot_args$orig_mod_id
+
+  # Write out new dataset
+  data_boot_name <- glue("boot_{orig_mod_id}_{mod_name}.csv")
```

Comment:
Use hyphen separator rather than underscore; but maybe this should just be `{mod_name}.csv` to keep things simple? that way the data name matches the control stream and yaml stems.